program/src/processor/deactivate_stake.rs

```diff
@@ -1,21 +1,136 @@
-use solana_program::{entrypoint::ProgramResult, pubkey::Pubkey};
+use std::num::NonZeroU64;
 
-use crate::instruction::accounts::{Context, DeactivateStakeAccounts};
+use solana_program::{
+    clock::Clock, entrypoint::ProgramResult, program_error::ProgramError, pubkey::Pubkey,
+    sysvar::Sysvar,
+};
+
+use crate::{
+    error::StakeError,
+    instruction::accounts::{Context, DeactivateStakeAccounts},
+    require,
+    state::{find_stake_pda, Config, Stake},
+};
+
+/// Helper to calculate the maximum amount that can be deactivated.
+#[inline(always)]
+fn get_max_deactivation_amount(total_amount: u64, basis_points: u16) -> Result<u64, ProgramError> {
+    total_amount
+        .checked_mul(basis_points as u64)
+        .ok_or(ProgramError::ArithmeticOverflow)?
+        .checked_div(10_000)
+        .ok_or(ProgramError::ArithmeticOverflow)
```

Comment:
nit: Just a tad cleaner, if you want!

```suggestion
    total_amount
        .checked_mul(basis_points as u64)
        .and_then(|p| p.checked_div(10_000))
        .ok_or(ProgramError::ArithmeticOverflow)
```

Reply:
Nice!